search: Add is:followed filter.

kennethnrk opened the pull request:
Create the is:followed search operator.
Fetch all messages that are from followed topics
using exists.
Update API documentation and changelog.

Co-authored-by: Nimish <medatwalnimish@gmail.com>

Fixes #27309.

Added `Followed` enum `3` to `visibility_policy` of `user_topics` endpoint in `zulip.yaml` .

Screenshots.

Top of view
![image](https://github.com/zulip/zulip/assets/2746074/04bdeda3-c240-454b-afe9-22afcb69a9b3)


Search suggestions
![image](https://github.com/zulip/zulip/assets/29176437/3299f29a-22ed-4e2a-9d39-7d142a5974e4)

Help menu.
![image](https://github.com/zulip/zulip/assets/29176437/31a72257-df0f-4c31-bbb6-1055a4f0e7ca)
![image](https://github.com/zulip/zulip/assets/29176437/9a1de00a-8072-4d38-9dad-4bb029997e93)

Api Change log
![image](https://github.com/zulip/zulip/assets/29176437/933785a3-de3f-47c2-832b-6c91ea798a46)
construct-narrow documentation
![image](https://github.com/zulip/zulip/assets/29176437/b5e680f3-5a69-4b90-a7e0-74ea3d5c0074)


<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

alya commented:
Thanks! To prepare this PR for review, please include screenshots of the search (with typeahead) and help center changes.

kennethnrk commented:
> Thanks! To prepare this PR for review, please include screenshots of the search (with typeahead) and help center / help menu changes.

Done

alya commented:
In the typeahead, let's make it "Followed topics", not "Topics followed".

alya commented:
https://zulip.com/help/search-for-messages#search-by-message-status needs to be updated as well.

kennethnrk commented:
> In the typeahead, let's make it "Followed topics", not "Topics followed".

Done

alya commented:
I'm not sure need to explicitly document `-is:followed` in the help center (we usually don't, except for unresolved topics, which are potentially of more interest than resolved topics). But other reviewers can comment on this as well.

alya commented:
@laurynmm Would it make sense for you to review this one? I haven't tested.

laurynmm commented on zerver/tests/test_message_fetch.py:
```diff
@@ -4533,6 +4563,37 @@ def test_has_reaction(self) -> None:
 
 
+class MessageIsTest(ZulipTestCase):
+    def test_message_is_followed(self) -> None:
```
Cordelia and Hamlet automatically follow topics when they send a message, so one could imagine making this test a bit more robust by having some followed topics and some not and testing as such. See the is resolved test for an example.

Also, as a prep commit, it might make sense to move that is resolved test to this new test class?

kennethnrk replied:
> Cordelia and Hamlet automatically follow topics when they send a message.

When I send a message from hamlet or cordelia there is no entry in the `UserTopic` table indicating that they've followed the topic. Is there some other table that indicates if a user is following a topic other than `UserTopic`?

laurynmm replied:
Ah, I was misreading the comments in `test_user_topics.AutomaticallyFollowTopicsTests`, but you could maybe do something similar here. If nothing else, I'd confirm via a test that a message in a topic that's not followed isn't returned.

laurynmm reviewed (commented):
@kennethnrk - Thanks for working on adding this narrow/search filter!

I left some comments below. Let me know if you have questions.

I did have one bigger question after doing some manual testing.

I created a new user in the dev environment and then I followed a topic in a stream I was subscribed to automatically (e.g. support) and then I subscribed to another stream (e.g. Denmark) and followed a topic in that stream (that existed before I subscribed). 

When I filter for followed topics, then only the messages in the automatically subscribed stream show up. And if I add the `channels:public` filter, then it's still only showing the messages I have UserMessage rows for and not all the messages in the topic.

I don't think this is the behavior we want? Instead, I would expect for all the messages in the followed topic to be returned in that second search. But then again, I get the same kind of results when I search for `channels:public is:resolved`. :slightly_frowning_face: 

<details>
<summary>Screenshot of is:followed and channels:public search</summary>

![Screenshot from 2024-04-15 16-00-51](https://github.com/zulip/zulip/assets/63245456/4a0b0c3e-b352-432b-b009-4479daace632)
</details>
<details>
<summary>Screenshot of the followed topic with messages before I subscribed</summary>

![Screenshot from 2024-04-15 16-01-03](https://github.com/zulip/zulip/assets/63245456/61d2fd1c-e153-4ef0-adb2-31af9b5292c5)
</details>
<details>
<summary>Screenshot of is:resolved and channels:public search</summary>

![Screenshot from 2024-04-15 16-30-07](https://github.com/zulip/zulip/assets/63245456/e8f53589-6cb9-4173-97b9-844bd490146d)
</details>
<details>
<summary>Screenshot of the resolved topic with messages before I subscribed</summary>

![Screenshot from 2024-04-15 16-30-16](https://github.com/zulip/zulip/assets/63245456/89d150e2-ef09-4c48-8df8-d42e581fa6ee)
</details>

Maybe it's not a big deal since, in most cases, a user is going to follow a topic that they either sent a message to or received a message in. They likely won't go back and follow topics/conversations that happened before they subscribed. Still it might be good to bring this up in a [#api design](https://chat.zulip.org/#narrow/stream/378-api-design) conversation just to clarify what the expectations are here.

If nothing else, it might make sense for the web app to not suggest searching in all public streams for these narrows as it won't provide any further search results. And also it would be good to note in searching in shared history help section that these filters do not return expanded results. But again, this needs to be a larger discussion on CZO before making those changes.

kennethnrk commented:
I have addressed the comments and made the required changes.

Opened a [CZO topic](https://chat.zulip.org/#narrow/stream/378-api-design/topic/is.3Aresolved.20operator.20only.20showing.20UserMessage.20rows) to discuss the question about `UserMessage` rows.

laurynmm commented on zerver/openapi/zulip.yaml:
```diff
@@ -6189,7 +6189,10 @@ paths:
 
-            **Changes**: In Zulip 9.0 (feature level 250), narrows gained support
+            **Changes**: In Zulip 9.0 (feature level 254), added new `is:followed`
+            filter, matching messages from followed topics.
```
I'd revise to something like: `filter, which returns messages in a user's [followed topics](help-link).`

It's always good to link to the help center article that explains a feature.

kennethnrk commented on api_docs/construct-narrow.md:
```diff
@@ -51,7 +51,11 @@ important optimization when fetching messages in certain cases (e.g.
-**Changes**: In Zulip 9.0 (feature level 250), support was added for
+**Changes**: In Zulip 9.0 (feature level 254), narrows gained support
+for a new filter `is:followed`. This allows clients to retrieve
+messages only from topics they have followed.
```
Ok, I have revised it to "This operator retrieves messages only
from topics that have been [followed](/help/follow-a-topic)."

kennethnrk commented:
> Also, it's good to post screenshots of any documentation changes/updates in your pull request comments. See [tips and best practices](https://zulip.readthedocs.io/en/latest/contributing/reviewable-prs.html#tips-and-best-practices) for pull requests.

Ok, I have addressed the comments and added the screenshots to the description.

kennethnrk commented:
Resolved conflicts

laurynmm commented:
@kennethnrk - Ping me for a review when you've got the merge conflicts resolved.

And I think we still need to get clarity in the [CZO discussion you started](https://chat.zulip.org/#narrow/stream/9-issues/topic/is.3Aresolved.20operator.20only.20showing.20UserMessage.20rows/near/1783979) about a `streams: public` and `is:followed` search. Because currently, this implementation only returns messages in followed topics with UserMessage rows, correct?

kennethnrk reviewed (commented):
> @kennethnrk - Ping me for a review when you've got the merge conflicts resolved.

Done. @laurynmm 

> Because currently, this implementation only returns messages in followed topics with UserMessage rows, correct? 

Yes, that is correct. Most likely something to do with the base query?

laurynmm reviewed (commented):
@kennethnrk - Thanks for the ping! I did a review pass focusing mostly on the documentation changes. Let me know if you have any questions about my comments.

In addition to those notes:
- Fixing the query to work correctly when combined with the `channels:public` filter should happen. Figuring out what the issue is with `is:resolved` might be helpful to work out what needs to be updated for this new filter's implementation.
- The commit message here should be updated to follow our project [guidelines](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html#commit-messages).

Appreciate all your work on this!

kennethnrk commented:
I've addressed the comments, I have opened #30125 to fix the issue with `channel:public is:resolved`.

timabbott commented on api_docs/changelog.md:
```diff
@@ -20,6 +20,16 @@ format used by the Zulip server that they are interacting with.
+  [`POST /register`](/api/register-queue):
+  Added support for a [search/narrow filter](/api/construct-narrow)
+  related to followed topics: `is:followed`. This operator narrows to
+  messages from topics that the user has [followed](/help/follow-a-topic).
```
I kinda wonder if this should just link to https://zulip.com/api/construct-narrow rather than the individual endpoints, since that's the part of the API that you'd be looking at to see what's changed. Probably worth an "api documentation" stream discussion, since we'd want to change the existing ones if so.

kennethnrk replied:
I've opened a discussion at [CZO](https://chat.zulip.org/#narrow/stream/412-api-documentation/topic/Narrow.20filter.20documentation).

timabbott commented on api_docs/construct-narrow.md:
```diff
@@ -51,8 +51,12 @@ important optimization when fetching messages in certain cases (e.g.
-two filters related to channel messages: `channel` and `channels`. The
+**Changes**: In Zulip 9.0 (feature level 263), support was added for a
+filter related to followed topics: `is:followed`. This operator narrows to
+messages from topics that the user has [followed](/help/follow-a-topic).
```
Change "messages from topics" => "message in topics"; that's more precise. This is also overly wordy; I think we could just say:

> **Changes**: In Zulip 9.0 (feature level 263), support was added for a
new `is:followed` filter, matching messages in topics that the current user is
[following](/help/follow-a-topic).

timabbott commented on zerver/lib/topic_sqlalchemy.py:
```diff
@@ -18,3 +19,22 @@ def get_resolved_topic_condition_sa() -> ColumnElement[Boolean]:
+                literal_column("zerver_usertopic.visibility_policy")
+                == literal(UserTopic.VisibilityPolicy.FOLLOWED),
+                literal_column("zerver_usertopic.topic_name")
+                == literal_column("zerver_message.subject"),
```
This check is wrong -- topics are case-insensitive.

timabbott replied:
`topic_match_sa` in this same file has correct logic you can use.

Can you also post the SQL that this generates, and its query plan from `EXPLAIN ANALYZE`? We may need to adjust some indexes to make this efficient.

kennethnrk replied:
> This check is wrong -- topics are case-insensitive.

Ok, a related issue that I noticed:
If I try to create a new topic with the same name but different case, duplicate topics show up in the sidebar.
All messages are sent to the original topic itself and if i collapse the sidebar and reopen, the duplicate topic disappears.
![image](https://github.com/zulip/zulip/assets/29176437/4bf09933-d2f4-4361-b3ee-0079debcf2ce)

kennethnrk replied:
Query
```
SELECT message_id, flags
FROM zerver_usermessage
JOIN zerver_message ON zerver_usermessage.message_id = zerver_message.id
WHERE user_profile_id = 33 
AND EXISTS (
    SELECT 1 
    FROM zerver_usertopic 
    WHERE zerver_usertopic.user_profile_id = 33 
    AND zerver_usertopic.visibility_policy = 3 
AND upper(zerver_usertopic.topic_name) = upper(zerver_message.subject)
    AND zerver_usertopic.recipient_id = zerver_message.recipient_id
) 
ORDER BY message_id DESC
LIMIT 50;

```
EXPLAIN ANALYZE

```
                                                                                             QUERY PLAN

-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=265.27..265.28 rows=1 width=12) (actual time=1.442..1.449 rows=50 loops=1)
   ->  Sort  (cost=265.27..265.28 rows=1 width=12) (actual time=1.440..1.443 rows=50 loops=1)
         Sort Key: zerver_usermessage.message_id DESC
         Sort Method: top-N heapsort  Memory: 29kB
         ->  Nested Loop  (cost=256.96..265.26 rows=1 width=12) (actual time=0.311..1.399 rows=170 loops=1)
               ->  Nested Loop  (cost=256.67..264.45 rows=1 width=4) (actual time=0.301..0.964 rows=170 loops=1)
                     ->  HashAggregate  (cost=1.00..1.01 rows=1 width=142) (actual time=0.023..0.026 rows=3 loops=1)
                           Group Key: upper((zerver_usertopic.topic_name)::text), zerver_usertopic.recipient_id
                           Batches: 1  Memory Usage: 24kB
                           ->  Seq Scan on zerver_usertopic  (cost=0.00..1.00 rows=1 width=142) (actual time=0.015..0.017 rows=3 loops=1)
                                 Filter: ((user_profile_id = 30) AND (visibility_policy = 3))
                                 Rows Removed by Filter: 2
                     ->  Bitmap Heap Scan on zerver_message  (cost=255.66..263.42 rows=2 width=27) (actual time=0.200..0.305 rows=57 loops=3)
                           Recheck Cond: ((recipient_id = zerver_usertopic.recipient_id) AND (upper((subject)::text) = upper((zerver_usertopic.topic_name)::text)))
                           Heap Blocks: exact=157
                           ->  BitmapAnd  (cost=255.66..255.66 rows=2 width=0) (actual time=0.184..0.184 rows=0 loops=3)
                                 ->  Bitmap Index Scan on zerver_message_recipient_id_5a7b6f03  (cost=0.00..5.62 rows=179 width=0) (actual time=0.013..0.013 rows=221 loops=3)
                                       Index Cond: (recipient_id = zerver_usertopic.recipient_id)
                                 ->  Bitmap Index Scan on zerver_message_realm_upper_subject  (cost=0.00..249.78 rows=52 width=0) (actual time=0.160..0.160 rows=57 loops=3)
                                       Index Cond: (upper((subject)::text) = upper((zerver_usertopic.topic_name)::text))
               ->  Index Scan using zerver_usermessage_user_profile_id_message_id_4936d0df_uniq on zerver_usermessage  (cost=0.29..0.81 rows=1 width=12) (actual time=0.002..0.002 rows=1 loops=170)
                     Index Cond: ((user_profile_id = 30) AND (message_id = zerver_message.id))
 Planning Time: 0.901 ms
 Execution Time: 1.578 ms
(24 rows)
```

timabbott replied:
One should always cross-link threads in both directions. This was:
https://chat.zulip.org/#narrow/stream/3-backend/topic/is.3Afollowed.20operator/near/1820472

Where we decided the query is probably fine as written.

timabbott commented on zerver/lib/narrow_predicate.py:
```diff
@@ -66,6 +66,8 @@ def satisfies_operator(*, operator: str, operand: str) -> bool:
                 if not topic_name.startswith(RESOLVED_TOPIC_PREFIX):
                     return False
+            elif operator == "is" and operand == "followed" and message["type"] != "stream":
+                return False
```
This seems to not actually check if the message is in a topic that the current user follows.

kennethnrk replied:
This logic was written by @nimishmedatwal. 

From what I understand, `build_narrow_predicate` is used in `tornado` to decide which event updates to send to the client. We're not passing the user profile in this function so it'll require a lot of changes to actually check if the user is following a topic.

By default, all events for new messages will be sent. This check filters out only the events with `type == stream` and only sends those events in the case of `is followed`.

So, this means that even though this check is incorrect, it will not produce undesired results.

kennethnrk replied:
On further testing, I've noticed that no matter what narrow the user is on, the corresponding `narrow` in `build_narrow_predicate` and `ClientDescriptor` in `event_queue.py` which calls `build_narrow_predicate` is always `[ ]`.

This seems like a bug? Or is there something I'm missing?

timabbott replied:
This refers to a part of the API for a single-narrow Zulip feed. 

https://zulip.com/api/register-queue#parameter-narrow

You can use the example API client [here](https://github.com/zulip/python-zulip-api/blob/main/zulip/zulip/examples/get-messages) to mess with it. I expect it to be somewhat rare for someone to use this API with anything but stream/channel narrows. But I think the correct answer is that this function should be rejecting this particular operator, because the `build_narrow_predicate` function cannot be correctly implemented without access to the database/user, like the other unsupported operators.

```
def check_narrow_for_events(narrow: Collection[NarrowTerm]) -> None:                                    
    supported_operators = [*channel_operators, "topic", "sender", "is"]                                 
    for narrow_term in narrow:                                                                          
        operator = narrow_term.operator                                                                 
        if operator not in supported_operators:                                                         
            raise JsonableError(_("Operator {operator} not supported.").format(operator=operator))      
```

kennethnrk replied:
Ok, I've modified `check_narrow_for_events` to reject `is:followed`.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -7071,6 +7093,24 @@ paths:
+                    In Zulip 7.0 (feature level 177), narrows gained support
+                    for three new filters related to direct messages: `is:dm`, `dm` and
+                    `dm-including`; replacing and deprecating `is:private`, `pm-with` and
+                    `group-pm-with` respectively.
```
These changes look like a misresolved merge conflict. It's clear that we intend for this documentation to have moved, see `**Changes**: See changelog at [Construct a narrow](/api/construct-narrow).`

In general, whenever you see a merge conflict, you'll want to use `git log -p <path>` to read the change that your work conflicted with and understand how to make your changes correctly on top of those. Doing such a workflow should make this sort of failure mode where you accidentally undo someone else's work very unlikely.

timabbott commented on zerver/tests/test_message_fetch.py:
```diff
@@ -4584,6 +4614,40 @@ def test_has_reaction(self) -> None:
+            user_profile=self.example_user("iago"),
+            visibility_policy=UserTopic.VisibilityPolicy.FOLLOWED,
+        )
+        my_topic.save()
```
Please use the API to follow the topic, rather than just creating database objects. Doing it this way is very fragile, and tends to break if refactoring occurs.

timabbott commented:
I did a pretty complete review on this one from reading it, but haven't done any manual testing. Probably Alya can do a round of that once you've addressed the potential bugs, since the most likely things for us to find will be strings we want to change.

kennethnrk commented:
While using the API for testing, I found that the `followed` enum was missing for the `user_topics` endpoint in the open api yaml file. So I added that as a pre-commit.

timabbott commented:
Ask for a review from Alya once the remaining details discussed here are resolved.

alya commented:
Cool! Since we've been adding descriptions to the top nav bar, let's do that here as well:

> Messages in topics you follow [?]

The [?] link should go to `/help/follow-a-topic`.

alya commented:
Looks good to me otherwise! Please mark the PR for integration review once that's been addressed.

kennethnrk commented:
Thanks @alya. I've made the header changes and added the screenshot to the description.

@zulipbot add "integration review"

kennethnrk commented:
@zulipbot remove "maintainer review"

alya commented:
Hm, I don't see the relevant screenshot?

timabbott commented on zerver/lib/narrow_predicate.py:
```diff
@@ -14,10 +14,15 @@
 
 def check_narrow_for_events(narrow: Collection[NarrowTerm]) -> None:
     supported_operators = [*channel_operators, "topic", "sender", "is"]
+    unsupported_operands = ["followed"]
```
This should be named `unsupported_is_operands`.

timabbott commented:
I added the screenshot that @alya had requested as "Top of view" above, which looks fine. Also made the change in https://github.com/zulip/zulip/pull/29667#discussion_r1643580898 and did some fixups to the documentation; the first hunk just simplifying the text and uses the consistent "messages in topics" and present tense to be precise -- it's whether you follow the topic right now, not anything related to when the message was sent. Diff below.

With those changes, this looks good to me, marked to merge once CI passes.

``` diff

diff --git a/api_docs/changelog.md b/api_docs/changelog.md
index 58953bb15d..3556be3e84 100644
--- a/api_docs/changelog.md
+++ b/api_docs/changelog.md
@@ -26,9 +26,9 @@ format used by the Zulip server that they are interacting with.
   [`GET /messages/matches_narrow`](/api/check-messages-match-narrow),
   [`POST /messages/flags/narrow`](/api/update-message-flags-for-narrow),
   [`POST /register`](/api/register-queue):
-  Added support for a [search/narrow filter](/api/construct-narrow)
-  related to followed topics: `is:followed`. This operator narrows to
-  messages from topics that the user has [followed](/help/follow-a-topic).
+  Added a new [search/narrow filter](/api/construct-narrow),
+  `is:followed`, matching messages in topics that the current user is
+  [following](/help/follow-a-topic).
 
 **Feature level 264**
 
diff --git a/api_docs/construct-narrow.md b/api_docs/construct-narrow.md
index e8eaf87390..0f69e7812f 100644
--- a/api_docs/construct-narrow.md
+++ b/api_docs/construct-narrow.md
@@ -51,7 +51,7 @@ important optimization when fetching messages in certain cases (e.g.
 when [adding the `read` flag to a user's personal
 messages](/api/update-message-flags-for-narrow)).
 
-**Changes**: In Zulip 9.0 (feature level 265),, support was added for a
+**Changes**: In Zulip 9.0 (feature level 265), support was added for a
 new `is:followed` filter, matching messages in topics that the current
 user is [following](/help/follow-a-topic).
 
```
